test/elixir_internal_certificate/account/accounts_test.exs

```diff
@@ -0,0 +1,204 @@
+defmodule ElixirInternalCertificate.Account.AccountsTest do
+  use ElixirInternalCertificate.DataCase
+
+  alias ElixirInternalCertificate.Account.Accounts
+  alias ElixirInternalCertificate.Account.Schemas.{User, UserToken}
+
+  describe "get_user_by_email/1" do
+    setup do
+      %{user: insert(:user)}
+    end
+
+    test "when the email does not exist, does not return the user" do
+      assert Accounts.get_user_by_email("unknown@example.com") == nil
+    end
+
+    test "with the existed email, returns the user", %{user: user} do
```

Comment:
when we read this test, we need to scroll up to see where does the user come from, it makes the test hard to read and maintain, as I explained last week, if possible, we tend to make each test case should be

- isolate
- easy to read, understand and fix

Reading the test isn't reading the codebase, we should make it as simplest as possible, that principle will help us when our project grows big where we have thousands of tests and our codebase is years old, maintaining the test will be really hard if we don't organize it well, so why not just keep everything as simplest as possible at the beginning.

Let's do the comparison between the 2 approaches

```
describe "some_feature/1" do
    setup do
      %{user: insert(:user)}   
    end

    # a few test cases

    test "with the existed email, returns the user", %{user: user} do
       %{id: id} = user

      assert %User{id: ^id} = Accounts.get_user_by_email(user.email)
    end
```

vs 

```
describe "some_feature/1" do
    # a few test cases

    test "with the existed email, returns the user" do
       %{id: id} = insert(:user, email: "aaa@gmail.com")

      assert %User{id: ^id} = Accounts.get_user_by_email("aaa@gmail.com")
    end
```

Which style do you think it's easy to read, we do not need to think about the data relative topic? I guess the second one

You might find something useful in https://nimblehq.co/compass/development/web/testing/#general